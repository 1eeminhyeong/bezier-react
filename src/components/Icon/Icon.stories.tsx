/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import { Meta, Story } from '@storybook/react'

/* Internal dependencies */
import { styled } from 'Foundation'
import { getObjectFromEnum, getTitle, iconList } from 'Utils/storyUtils'
import { Text } from 'Components/Text'
import Icon from './LegacyIcon'
import IconProps, { IconSize } from './Icon.types'

export default {
  title: getTitle(base),
  component: Icon,
  argTypes: {
    size: {
      control: {
        type: 'radio',
        options: getObjectFromEnum(IconSize),
      },
    },
  },
} as Meta

const IconInfo = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 120px;
  height: 120px;
`

const Name = styled.p`
  text-align: center;
`

export const AllIcons = (args) => (
  <>
    { iconList.map((iconName) => (
      <IconInfo key={iconName}>
        <Icon
          name={iconName}
          {...args}
        />
        <Name>{ iconName }</Name>
      </IconInfo>
    )) }
  </>
)
AllIcons.args = {
  size: IconSize.Normal,
  color: 'bgtxt-olive-dark',
}

const Template: Story<IconProps> = (args) => <Icon {...args} />

export const Primary = Template.bind({})
Primary.args = {
  name: 'channel',
  color: 'bgtxt-olive-dark',
  size: IconSize.Normal,
  marginTop: 0,
  marginRight: 0,
  marginBottom: 0,
  marginLeft: 0,
}

export const WithText: Story<IconProps> = ({
  color,
  name,
  ...otherIconProps
}) => (
  <Text style={{ color }}>
    <Icon
      name={name}
      {...otherIconProps}
    />
    Hello World!
  </Text>
)
WithText.args = {
  name: 'channel',
  color: 'bgtxt-olive-dark',
  size: IconSize.Normal,
}
