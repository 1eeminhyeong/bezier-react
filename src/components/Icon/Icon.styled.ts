/* Internal dependencies */
import { styled, css, SemanticNames } from '../../foundation'

interface IconStyledProps {
  color?: SemanticNames
  hoverColor?: SemanticNames
  margintop: number
  marginright: number
  marginbottom: number
  marginleft: number
}

function getMargin({
  margintop,
  marginright,
  marginbottom,
  marginleft,
}: IconStyledProps): string {
  return `${margintop}px ${marginright}px ${marginbottom}px ${marginleft}px`
}

const Icon = styled.svg<IconStyledProps>`
  flex: 0 0 auto;
  margin: ${getMargin};
  color: ${({ foundation, color }) => (color && foundation?.theme?.[color]) || 'inherit'};

  ${({ foundation }) => foundation?.transition?.getTransitionsCSS('color')}

  ${({ foundation, hoverColor }) => hoverColor && css`
    &:hover {
      color: ${foundation?.theme?.[hoverColor]};
    }
  `}
`

export default Icon
