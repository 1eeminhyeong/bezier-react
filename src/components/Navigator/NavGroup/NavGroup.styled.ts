/* Internal dependencies */
import { styled, css, ellipsis } from 'Foundation'
import type { InterpolationProps } from 'Types/Foundation'
import { Text } from 'Components/Text'

interface WrapperProps extends InterpolationProps {
  open: boolean
}

export const LeftIconWrapper = styled.div`
  display: flex;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
`

export const ContentWrapper = styled(Text)`
  ${ellipsis()}
`

export const ChevronWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: 2px;
`

export const RightContentWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
`

const closedItemStyle = css`
  ${ChevronWrapper} {
    visibility: hidden;
  }

  &:hover,
  &:focus-visible {
    ${ChevronWrapper} {
      visibility: visible;
    }
  }
`

export const Item = styled.button<WrapperProps>`
  display: flex;
  align-items: center;
  width: 100%;
  height: 28px;
  padding: 0 6px;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  outline: 0;
  ${({ foundation }) => foundation?.rounding?.round6}

  &:hover,
  &:focus-visible {
    background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
  }

  ${({ open }) => (open ? null : closedItemStyle)}

  ${({ interpolation }) => interpolation}
`

export const ChildrenWrapper = styled.ul`
  padding: 0 0 8px;
  margin: 0;
`

export const Wrapper = styled.li`
  display: block;
`
