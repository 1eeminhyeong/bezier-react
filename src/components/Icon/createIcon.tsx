/* External dependencies */
import React from 'react'

/* Internal dependencies */
import Icon from './Icon'
import type { IconProps } from './Icon.types'

function createIcon(iconFn: (props: React.SVGProps<SVGSVGElement>) => JSX.Element) {
  return ((props: IconProps) => (
    <Icon
      as={iconFn}
      {...props}
    />
  ))
}

export default createIcon
