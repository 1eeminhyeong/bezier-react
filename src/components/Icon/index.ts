/* Internal dependencies */
import Icon from './LegacyIcon'
import { IconSize } from './Icon.types'
import { isIconName } from './utils'
import type IconProps from './Icon.types'

export * from './generated'

export type { IconProps }

export {
  Icon,
  IconSize,
  isIconName,
}
