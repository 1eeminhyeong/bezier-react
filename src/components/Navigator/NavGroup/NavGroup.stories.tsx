/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from 'Utils/storyUtils'
import { Icon, IconSize } from 'Components/Icon'
import { NavItem } from 'Components/Navigator/NavItem'
import NavGroup from './NavGroup'
import type NavGroupProps from './NavGroup.types'

export default {
  title: getTitle(base),
  component: NavGroup,
} as Meta

const Template: Story<NavGroupProps> = (args) => (
  <div style={{ width: '240px', padding: '6px' }}>
    <NavGroup {...args}>
      <NavItem
        name="profile"
        content="채널 프로필"
      />
      <NavItem
        name="btn"
        content="버튼 설치 및 설정"
      />
    </NavGroup>
  </div>
)

export const Primary: Story<NavGroupProps> = Template.bind({})

Primary.args = {
  open: true,
  name: 'general',
  content: '일반 설정',
  leftIcon: 'settings',
  leftIconColor: 'txt-black-dark',
  rightContent: <Icon name="dot" size={IconSize.XS} color="bgtxt-orange-normal" />,
}
