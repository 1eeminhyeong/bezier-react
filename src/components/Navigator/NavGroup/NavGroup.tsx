/* External dependencies */
import React, { useCallback, memo } from 'react'
import { noop, isNil } from 'lodash-es'

/* Internal dependencies */
import { Typography } from 'Foundation'
import { Icon, IconSize, isIconName } from 'Components/Icon'
import type NavGroupProps from './NavGroup.types'
import {
  Item,
  Wrapper,
  ChildrenWrapper,
  LeftIconWrapper,
  ChevronWrapper,
  ContentWrapper,
  RightContentWrapper,
} from './NavGroup.styled'

export const NAV_GROUP_TEST_ID = 'bezier-react-nav-group'

function NavGroup({
  as,
  testId = NAV_GROUP_TEST_ID,
  name,
  style,
  className,
  interpolation,
  children,
  content,
  rightContent,
  leftIcon,
  leftIconColor = 'txt-black-dark',
  open,
  active,
  onClick = noop,
}: NavGroupProps) {
  const handleClickItem = useCallback((e?: React.MouseEvent) => {
    onClick(e, name)
  }, [
    name,
    onClick,
  ])

  const hasChildren = !isNil(children)
  const chevronIconName = open ? 'chevron-small-down' : 'chevron-small-right'
  const showLeftIcon = isIconName(leftIcon)
  const ariaName = `${name}Menu`

  return (
    <Wrapper role="none">
      <Item
        as={as}
        active={active}
        style={style}
        className={className}
        interpolation={interpolation}
        onClick={handleClickItem}
        data-testid={testId}
        role="menuitem"
        aria-haspopup={hasChildren}
        aria-expanded={open}
        aria-controls={ariaName}
      >
        <LeftIconWrapper>
          { showLeftIcon && (
            <Icon
              name={leftIcon}
              size={IconSize.S}
              color={leftIconColor}
            />
          ) }
        </LeftIconWrapper>

        <ContentWrapper typo={Typography.Size14}>
          { content }
        </ContentWrapper>

        { hasChildren && (
          <ChevronWrapper>
            <Icon
              name={chevronIconName}
              size={IconSize.S}
              color="txt-black-dark"
            />
          </ChevronWrapper>
        ) }

        { rightContent && (
          <RightContentWrapper>
            { rightContent }
          </RightContentWrapper>
        ) }
      </Item>

      { open && (
        <ChildrenWrapper
          role="menu"
          id={ariaName}
        >
          { open && children }
        </ChildrenWrapper>
      ) }
    </Wrapper>
  )
}

export default memo(NavGroup)
