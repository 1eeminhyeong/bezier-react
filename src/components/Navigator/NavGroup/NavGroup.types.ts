/* Internal dependencies */
import type {
  BezierComponentProps,
  ChildrenProps,
  ContentProps,
  SideContentProps,
  ActivatableProps,
  AdditionalColorProps,
} from 'Types/ComponentProps'
import { IconName } from 'Components/Icon'

interface NavGroupOptions {
  open?: boolean
  leftIcon: IconName
  name: string
  onClick?: (e?: React.MouseEvent, name?: string) => void
}

export default interface NavGroupProps extends
  BezierComponentProps,
  ChildrenProps,
  ContentProps,
  Pick<SideContentProps, 'rightContent'>,
  AdditionalColorProps<'leftIcon'>,
  Pick<ActivatableProps, 'active'>,
  NavGroupOptions {}
