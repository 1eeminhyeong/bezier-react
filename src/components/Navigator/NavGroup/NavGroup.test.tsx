/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { render } from 'Utils/testUtils'
import { Icon, IconSize } from 'Components/Icon'
import NavGroup, { NAV_GROUP_TEST_ID } from './NavGroup'
import type NavGroupProps from './NavGroup.types'

describe('NavGroup Test >', () => {
  let props: NavGroupProps

  beforeEach(() => {
    props = {
      leftIcon: 'dot',
      name: 'general',
      content: 'test-content',
      rightContent: <Icon name="dot" size={IconSize.XS} color="bgtxt-orange-normal" />,
    }
  })

  const renderNavItem = (optionProps?: Partial<NavGroupProps>) =>
    render(<NavGroup {...props} {...optionProps} />)

  it('Snapshot > active', () => {
    const { getByTestId } = renderNavItem({ active: true })

    const rendered = getByTestId(NAV_GROUP_TEST_ID)

    expect(rendered).toMatchSnapshot()
  })

  it('Snapshot > not active', () => {
    const { getByTestId } = renderNavItem({ active: false })

    const rendered = getByTestId(NAV_GROUP_TEST_ID)

    expect(rendered).toMatchSnapshot()
  })
})
