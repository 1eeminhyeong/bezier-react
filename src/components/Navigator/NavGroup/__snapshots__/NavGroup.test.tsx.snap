// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem Test > Snapshot > active 1`] = `
.c2 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c6 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #F4800B;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c3 {
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: inherit;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-left: auto;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 28px;
  padding: 0 6px;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  outline: 0;
  overflow: hidden;
  border-radius: 6px;
  color: #5E56F0;
  background-color: #5E56F01A;
}

.c0 .sc-eCApnc {
  visibility: hidden;
}

.c0:hover .sc-eCApnc,
.c0:focus-visible .sc-eCApnc {
  visibility: visible;
}

<button
  aria-controls="generalMenu"
  aria-haspopup="false"
  class="c0"
  data-testid="bezier-react-nav-group"
  role="menuitem"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      color="txt-black-dark"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="20"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="c3 c4"
    data-testid="bezier-react-text"
  >
    test-content
  </span>
  <div
    class="c5"
  >
    <svg
      class="c6"
      color="bgtxt-orange-normal"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="16"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</button>
`;

exports[`NavItem Test > Snapshot > not active 1`] = `
.c2 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c6 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #F4800B;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c3 {
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: inherit;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-left: auto;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 28px;
  padding: 0 6px;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  outline: 0;
  overflow: hidden;
  border-radius: 6px;
  color: #000000D9;
}

.c0:hover {
  background-color: #0000000D;
}

.c0:focus-visible {
  background-color: #0000000D;
}

.c0 .sc-eCApnc {
  visibility: hidden;
}

.c0:hover .sc-eCApnc,
.c0:focus-visible .sc-eCApnc {
  visibility: visible;
}

<button
  aria-controls="generalMenu"
  aria-haspopup="false"
  class="c0"
  data-testid="bezier-react-nav-group"
  role="menuitem"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      color="txt-black-dark"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="20"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="c3 c4"
    data-testid="bezier-react-text"
  >
    test-content
  </span>
  <div
    class="c5"
  >
    <svg
      class="c6"
      color="bgtxt-orange-normal"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="16"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</button>
`;
